assets/js/googlesitekit/widgets/datastore/areas.js

```diff
@@ -0,0 +1,215 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * External dependencies
+ */
+import invariant from 'invariant';
+
+/**
+ * Store our widget components by registry, then by widget `slug`. We do this because
+ * we can't store React components in our data store.
+ *
+ * @private
+ * @since n.e.x.t
+ */
+export const WidgetComponents = {};
+
+const ASSIGN_WIDGET_AREA = 'ASSIGN_WIDGET_AREA';
+const REGISTER_WIDGET_AREA = 'REGISTER_WIDGET_AREA';
+
+export const INITIAL_STATE = {
+	areas: {},
+	areaWidgets: {},
+	contexts: {
+		analytics: [],
+		dashboard: [],
+	},
+};
+
+export const actions = {
+	/**
+	 * Assigns a widget area to one (or several) contexts.
+	 *
+	 * Accepts an area slug to register as the first argument, then either a string
+	 * (for a single context slug) or array of contexts slugs (to assign the widget area
+	 * to multiple contexts).
+	 *
+	 * Does not error if any of the requested contexts do not exist, but will console.warn
+	 * a message about non-existant contexts.
+	 *
+	 * Does not error if the area slug provided does not exist, but will console.warn a
+	 * message about non-existent areas.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param  {string}         areaSlug     Widget Area's slug.
+	 * @param  {(string|Array)} contextSlugs Widget Context's slug(s).
+	 * @return {Object}                      Redux-style action.
+	 */
+	assignWidgetArea( areaSlug, contextSlugs ) {
+		let contextSlugsAsArray;
+		if ( typeof contextSlugs === 'string' ) {
+			contextSlugsAsArray = [ contextSlugs ];
+		} else {
+			contextSlugsAsArray = [ ...contextSlugsAsArray ];
+		}
+
+		return {
+			payload: { areaSlug, contextSlugs: contextSlugsAsArray },
+			type: ASSIGN_WIDGET_AREA,
+		};
+	},
+
+	/**
+	 * Creates a widget area with a given name and settings.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param  {string}             areaSlug           Widget Area's slug.
+	 * @param  {Object}             settings           Widget Area's settings.
+	 * @param  {number}             settings.priority  Priority for this widget area. Default: 10.
+	 * @param  {string}             settings.title     Title for this widget area.
+	 * @param  {string}             settings.subtitle  Subtitle for this widget area.
+	 * @param  {(string|undefined)} settings.icon      URL to SVG icon for this widget area.
+	 * @param  {string}             settings.style     Widget area style (one of "boxes", "composite").
+	 * @return {Object}                                Redux-style action.
+	 */
+	registerWidgetArea( areaSlug, settings ) {
+		invariant( areaSlug, 'areaSlug is required.' );
+		invariant( settings, 'settings is required.' );
+		invariant( settings.title, 'settings.title is required.' );
+		invariant( settings.subtitle, 'settings.subtitle is required.' );
+		invariant( settings.style, 'settings.style is required.' );
+
+		const settingsToUse = {
```

Comment:
Could we not just do `settings = { ...` and so on? `settings` is not a `const` and it wouldn't actually modify the input object itself since we don't modify that. It's not major, but I think it makes the code here a bit cleaner and also allows us to just pass `settings` in the payload below.

Reply:
It's generally not best practice in JS to mutate arguments passed to a function, as it can lead to unexpected things happening. In this case we aren't _entirely_ redefining it so it's not as bad, but I tend to avoid doing it entirely to avoid ambiguity.

I've refactored this a bit to use ES6 rest params, let me know what you think 😄